subsys/bluetooth/host/classic/sdp.c

```diff
@@ -2206,108 +2234,193 @@ static void sdp_client_connected(struct bt_l2cap_chan *chan)
 
 	LOG_DBG("session %p chan %p connected", session, chan);
 
+	k_sem_take(&session->sem_lock, K_FOREVER);
 	session->rec_buf = chan->ops->alloc_buf(chan);
 	if (!session->rec_buf) {
+		session->state = SDP_CLIENT_DISCONNECTING;
+		k_sem_give(&session->sem_lock);
 		bt_l2cap_chan_disconnect(chan);
 		return;
 	}
+	k_sem_give(&session->sem_lock);
 
 	sdp_client_discover(session);
 }
 
+static void sdp_client_clean_after_disconnect(struct bt_sdp_client *session)
+{
+	/*
+	 * keep the follow fields:
+	 * sem_lock - global, never clean it.
+	 * state - the seesion's state.
+	 * chan - it is still used before released callback.
+	 * reqs_next - the pending reqs in the disconnecting phase.
+	 */
+	(void)memset(&session->reqs, 0,
+		sizeof(*session) - offsetof(struct bt_sdp_client, reqs));
+}
+
+static void sdp_client_clean_after_release(struct bt_sdp_client *session)
+{
+	/*
+	 * keep the follow fields:
+	 * sem_lock - global, never clean it.
+	 */
+	/* session->state is set as SDP_CLIENT_RELEASED by memset */
+	(void)memset(&session->state, 0,
+		     sizeof(*session) - offsetof(struct bt_sdp_client, state));
```

Comment:
This still looks prone to bugs. E.g. if `session->recv_buf` is set it should be freed with `net_buf_unref()`. I'd propose just doing each member individually, and you can probably reuse one of these "clean()" functions from the other, i.e. `after_release()` can probably call `after_disconnect()` first to avoid duplicating too many lines of code.

Reply:
agree, updated.